### PR TITLE
Adding Static Maps API

### PR DESCRIPTION
**Pull request by domesticmouse**
Fixes https://github.com/googlemaps/google-maps-services-java/issues/233. 

Also contains random clean ups and fixes, e.g. https://github.com/googlemaps/google-maps-services-java/issues/412

### CODE REVIEW

File: src/main/java/com/google/maps/StaticMapsRequest.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    public void size(MarkersSize size) {
+      this.size = size;
+    }
+
+    /**
+     * Specifies a 24-bit color (example: color=0xFFFFCC) or a predefined color from the set {black,
+     * brown, green, purple, yellow, blue, gray, orange, red, white}.
+     */
+    public void color(String color) {
+      if (!colorPattern.matcher(color).matches()) {
```
Comment (samthor) [inline src/main/java/com/google/maps/StaticMapsRequest.java]: is it worth enforcing this locally? We might allow more colors in future.

File: src/main/java/com/google/maps/StaticMapsRequest.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        return this.name();
+      }
+    }
+
+    private MarkersSize size;
+    private String color;
+    private String label;
+    private String customIconURL;
+    private CustomIconAnchor anchorPoint;
+    private List<String> locations = new ArrayList<>();
```
Comment (samthor) [inline src/main/java/com/google/maps/StaticMapsRequest.java]: private final, you don't ever change this ref

File: src/main/java/com/google/maps/StaticMapsRequest.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    return param("markers", markers);
+  }
+
+  public static class Path implements UrlValue {
+
+    private Integer weight;
+    private String color;
+    private String fillcolor;
+    private boolean geodesic;
+    private List<String> points = new ArrayList<>();
```
Comment (samthor) [inline src/main/java/com/google/maps/StaticMapsRequest.java]: private final

File: src/main/java/com/google/maps/StaticMapsRequest.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+    /**
+     * Specifies the thickness of the path in pixels. If no weight parameter is set, the path will
+     * appear in its default thickness (5 pixels).
+     */
+    public void weight(int weight) {
+      this.weight = weight;
```
Comment (samthor) [inline src/main/java/com/google/maps/StaticMapsRequest.java]: should setting this to zero clear the Integer?

File: src/main/java/com/google/maps/model/Size.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+   *
+   * @param height The height of this Size.
+   * @param width The width of this Size.
+   */
+  public Size(int width, int height) {
+    this.width = width;
+    this.height = height;
+  }
+
+  /** Serialisation constructor. */
```
Comment (samthor) [inline src/main/java/com/google/maps/model/Size.java]: american spelling, s => z

File: src/test/java/com/google/maps/StaticMapsApiTest.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+import com.google.maps.StaticMapsRequest.Markers;
+import com.google.maps.StaticMapsRequest.Markers.CustomIconAnchor;
+import com.google.maps.StaticMapsRequest.Markers.MarkersSize;
+import com.google.maps.StaticMapsRequest.Path;
+import com.google.maps.StaticMapsRequest.StaticMapType;
+import com.google.maps.model.LatLng;
+import com.google.maps.model.Size;
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayInputStream;
+import javax.imageio.ImageIO;
```
Comment (samthor) [inline src/test/java/com/google/maps/StaticMapsApiTest.java]: Is this library only pulled in for testing? Do we have it for the regular lib?

File: src/main/java/com/google/maps/PendingResultBase.java [hunk 7]
```diff
@@ -96,8 +96,18 @@ protected A param(String key, String val) {
     return result;
   }
 
+  protected A param(String key, int val) {
+    params.put(key, Integer.toString(val));
+
+    @SuppressWarnings("unchecked") // safe by specification - A is the actual class of this instance
```
Comment (samthor) [inline src/main/java/com/google/maps/PendingResultBase.java]: minor nit: could you replace these three returns with

```java
private A getInstance() {
  @SuppressWarnings("unchecked");
  return (A) this;
}

// later
return this.getInstance();
```

Comment (domesticmouse) [reply src/main/java/com/google/maps/PendingResultBase.java]: I like it, done. Ish.

File: src/main/java/com/google/maps/ImageResult.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  public ImageResult(String contentType, byte[] imageData) {
+    this.imageData = imageData;
+    this.contentType = contentType;
+  }
+
+  private static final long serialVersionUID = 1L;
+
+  /** The image data from the Photos API call. */
+  public final byte[] imageData;
```
[the remark this answers is not in the document]
Comment (domesticmouse) [reply src/main/java/com/google/maps/ImageResult.java]: This `ImageResult` is an output value object. If a client of this library alters the image data, it has no impact on the library.

File: src/main/java/com/google/maps/StaticMapsRequest.java (hunk 2 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/StaticMapsRequest.java]: Done

File: src/main/java/com/google/maps/StaticMapsRequest.java (hunk 1 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/StaticMapsRequest.java]: Fair call, ripped out regex

File: src/main/java/com/google/maps/StaticMapsRequest.java (hunk 3 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/StaticMapsRequest.java]: done

File: src/main/java/com/google/maps/StaticMapsRequest.java (hunk 4 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/StaticMapsRequest.java]: Good idea. Implemented in `toUrlValue()`

File: src/main/java/com/google/maps/model/Size.java (hunk 5 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/model/Size.java]: done

File: src/test/java/com/google/maps/StaticMapsApiTest.java (hunk 6 above)
Comment (domesticmouse) [reply src/test/java/com/google/maps/StaticMapsApiTest.java]: I only use it in the test library to make sure we can round trip an image. I'm not decoding the image in the main library to keep up throughput.

File: src/main/java/com/google/maps/PendingResultBase.java [hunk 9]
```diff
@@ -89,32 +91,32 @@ public final void cancel() {
   protected abstract void validateRequest();
 
   protected A param(String key, String val) {
-    params.put(key, val);
+    if(params.get(key) == null) {
```
Comment (samthor) [inline src/main/java/com/google/maps/PendingResultBase.java]: nit: missing space `if (`

File: src/main/java/com/google/maps/PendingResultBase.java [hunk 10]
```diff
@@ -89,32 +91,32 @@ public final void cancel() {
...
+      params.put(key, new ArrayList<String>());
```
Comment (samthor) [inline src/main/java/com/google/maps/PendingResultBase.java]: This looks naïvely fine but I wonder if anyone is using this class to overwrite previous values being set via `.param()`.

Would it be worth having an explicit 'add' flag somehow? I know you already have the three `param` methods, so a matrix of those would be awkward. Let me know.

File: src/test/java/com/google/maps/LocalTestServerContext.java [hunk 11]
```diff
@@ -117,6 +117,21 @@ void assertParamValue(String expectedValue, String paramName)
...
+      this.params = this.actualParams();
+    }
+    int paramsFound = 0;
+    for (NameValuePair pair : params) {
+      if (pair.getName().equals(paramName)) {
+        assertEquals(expectedValues.get(paramsFound), pair.getValue());
+        paramsFound++;
+      }
+    }
+    assertTrue(paramsFound == expectedValues.size());
```
Comment (samthor) [inline src/test/java/com/google/maps/LocalTestServerContext.java]: why not assertEquals?

Comment (domesticmouse) [reply src/test/java/com/google/maps/LocalTestServerContext.java]: Good catch, fixed.

File: src/main/java/com/google/maps/PendingResultBase.java (hunk 10 above)
Comment (domesticmouse) [reply src/main/java/com/google/maps/PendingResultBase.java]: In fact, yes I was being naive. There was logic in optimise waypoints that implicitly relied on that underlying hashmap behavior of the parameters. 

New implementation now enforces singleton parameters (as implicitly required by non Static Maps API surfaces), and just uses the multiple parameters for Static Maps `path`s and `markers`.

Now the tests pass...

**Pull request merged**